src/modules/roc_core/target_linux/roc_core/timer.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2020 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <errno.h>
+#include <unistd.h>
+
+#include "roc_core/errno_to_str.h"
+#include "roc_core/panic.h"
+#include "roc_core/timer.h"
+
+namespace roc {
+namespace core {
+
+#if defined(CLOCK_MONOTONIC)
+#define CLOCK_TYPE CLOCK_MONOTONIC
+#else
+#define CLOCK_TYPE CLOCK_REALTIME
+#endif
+
+Timer::Timer()
+    : timerfd_(timerfd_create(CLOCK_TYPE, 0))
+    , deadline_(0) {
+    if (timerfd_ == -1) {
+        roc_panic("timer: timerfd_create(CLOCK_TYPE, 0): %s", errno_to_str().c_str());
+    }
+}
+
+Timer::~Timer() {
+    close(timerfd_);
+}
+
+bool Timer::try_set_deadline(nanoseconds_t new_deadline) {
+    if (!deadline_.try_store(new_deadline)) {
+        return false;
+    }
+
+    struct itimerspec new_value = convert_deadline();
+
+    int res = timerfd_settime(timerfd_, TFD_TIMER_ABSTIME, &new_value, NULL);
+    if (res == -1) {
+        roc_panic("timer: timerfd_settime(timerfd_, TFD_TIMER_ABSTIME, "
+                  "&new_value, NULL): %s",
```

Comment:
Nitpick: we usually omit non-interesting function arguments in log messages, so it can be:

```
timer: timerfd_settime(TFD_TIMER_ABSTIME): %s
```

or even just:

```
timer: timerfd_settime(): %s
```

Reply:
If we really can't rely that CLOCK_MONOTINIC is always available and should be able to work with CLOCK_REALTIME, then according to the man page we should be ready to get ECANCELED here: https://man7.org/linux/man-pages/man2/timerfd_settime.2.html#NOTES

Reply:
ECANCELED can get from read if TFD_TIMER_CANCEL_ON_SET flag is set. TFD_TIMER_CANCEL_ON_SET is off now.